### PR TITLE
(PUP-4420) Ensure pluginfacts don't ignore source permissions

### PR DESCRIPTION
**Pull request by whopper**
Prior to this commit, the :source_permissions attribute was
not being properly propagated into the Posix MetaStat class,
and so source permissions were ignored for all resources.

External facts are required to use source permissions so
that the facts are executable when they are downloaded onto
the agent. This means that due to this regression, external
facts were being pluginsync'd to the client without execution
permissions.

This commit re-enables using source permissions for fact
pluginsync, while continuing to ignore them otherwise.

See commit 21edcd6 for the first fix for of a similar, earlier
regression.

### CODE REVIEW

**Comment by whopper** [comment]
Putting this up now as I work on an acceptance test to make sure I'm headed in the right direction. @joshcooper given your fix for the exact same regression in June, does this look sane?

File: lib/puppet/type/file.rb [hunk 1]
```diff
@@ -668,6 +668,7 @@ def perform_recursion(path)
       :links => self[:links],
       :recurse => (self[:recurse] == :remote ? true : self[:recurse]),
       :recurselimit => self[:recurselimit],
+      :source_permissions => self[:source_permissions],
```
Comment (joshcooper) [inline lib/puppet/type/file.rb]: This looks correct to me. The agent needs to specify whether it wants to use them (or not), and the server should process the request based on the query parameters specified in the `file_metadatas` search request.

But this makes me wonder how it ever worked? I'm guessing the ruby implementation of the puppet file serving code mistakenly had a hardcoded assumption that `file_metadatas` search requests should always use source permissions?

File: lib/puppet/file_serving/terminus_helper.rb [hunk 2]
```diff
@@ -26,7 +26,8 @@ def path2instances(request, *paths)
     end
 
     Puppet::FileServing::Fileset.merge(*filesets).collect do |file, base_path|
-      path2instance(request, base_path, :ignore_source_permissions => true, :relative_path => file)
+      ignore_source_permissions = (!request.options[:source_permissions] || request.options[:source_permissions] == :ignore)
```
Comment (joshcooper) [inline lib/puppet/file_serving/terminus_helper.rb]: The Metadata class also has logic about how to interpret the default `source_permissions` value: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/file_serving/metadata.rb#L30. I'm thinking we should keep the logic in the Metadata class, and have the terminus_helper treat `source_permissions` like it does `checksum_type`, have the terminus_helper call the `#source_permissions=` method on the model class if specified in `request.options[:source_permissions]`.

I've pushed up a commit https://github.com/joshcooper/puppet/commit/07eb06f7f1517866314fab5459c1abef22959686 to describe what I mean. Amazingly only one test failed (because the test was passing an invalid source_permissions value).

Comment (whopper) [inline lib/puppet/file_serving/terminus_helper.rb]: @joshcooper cool, I'll see about integrating that into this PR!

File: lib/puppet/type/file.rb (hunk 1 above)
Comment (whopper) [inline lib/puppet/type/file.rb]: @joshcooper as far as I could tell, it was working previously because we only ignored source permissions if the option was explicitly set to `:ignore`, and since we weren't actually passing any options in, `source_permissions` was always `nil` (except when we actually were passing `:ignore` in).

`@source_permissions_ignore = source_permissions == :ignore      # resulted in :use when source_permissions was nil`

The change that caused the regression essentially broke this weird assumption that `nil` should use source permissions by saying that if `source_permissions` is `nil`, then we should ignore source permissions. This is why we now need the agent to specify `:use` if it actually needs to use source permissions. Anything that was depending on simply not setting `source_permissions` to `:use` them would be broken. I believe that's just external facts, however.

`@source_permissions_ignore = (!source_permissions || source_permissions == :ignore)`

File: lib/puppet/file_serving/metadata.rb [hunk 3]
```diff
@@ -22,10 +22,16 @@ def checksum_type=(type)
     @checksum_type = type
   end
 
+  def source_permissions=(source_permissions)
+    raise(ArgumentError, "Unsupported source_permission #{source_permissions}") unless [:use, :use_when_creating, :ignore].include?(source_permissions.intern)
```
Comment (whopper) [inline lib/puppet/file_serving/metadata.rb]: @joshcooper ran into some weirdness here. `source_permissions` was being compared to the symbol array as a string, and so Puppet was throwing exceptions.. `intern'd` it for now, but I'm not sure I understand why that's happening (I verified `source_permissions` is a symbol itself).

File: acceptance/tests/pluginsync/4420_pluginfacts_should_be_resolvable_on_agent.rb [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+master_opts = {
+  'main' => {
+    'environmentpath' => "#{codedir}/environments"
+  }
+}
+
+step "Pluginsync the external fact to the agent and ensure it resolves correctly"
+with_puppet_running_on master, master_opts, codedir do
+  agents.each do |agent|
+    on(agent, puppet('agent', "-t --server #{master}"), :acceptable_exit_codes => [2])
```
Comment (whopper) [inline acceptance/tests/pluginsync/4420_pluginfacts_should_be_resolvable_on_agent.rb]: Second note: the `notify` in the test manifest used here causes Puppet to exit with `2`, so I made that the acceptable exit code.

**Comment by whopper** [comment]
So far I've tested these changes and the acceptance test in linux by running against patched preserved hosts. Next up is Windows.

File: lib/puppet/file_serving/metadata.rb [hunk 5]
```diff
@@ -50,7 +56,7 @@ class WindowsStat < MetaStat
       require 'puppet/util/windows/security'
     end
 
-    def initialize(stat, path, source_permissions = nil)
```
Comment (whopper) [inline lib/puppet/file_serving/metadata.rb]: @joshcooper one windows spec test is failing because it was depending on `source_permissions` being `nil` by default here. Could that be a problem elsewhere, or do we expect to always pass something in? (If `nil` or `ignore` isn't explicitly passed in here, `@source_permissions_ignore` will not be set. Shouldn't it be true by default now?)

File: lib/puppet/file_serving/content.rb [hunk 6]
```diff
@@ -23,7 +23,7 @@ def self.from_binary(content)
 
   # This is no longer used, but is still called by the file server implementations when interacting
   # with their model abstraction.
-  def collect(source_permissions = nil)
+  def collect
```
Comment (joshcooper) [inline lib/puppet/file_serving/content.rb]: I probably shouldn't have done this, as it's a breaking change.

File: lib/puppet/file_serving/metadata.rb [hunk 7]
```diff
@@ -66,23 +72,23 @@ def initialize(stat, path, source_permissions = nil)
...
-      MetaStat.new(stat, source_permissions)
+      MetaStat.new(stat, @source_permissions)
     end
   end
 
   # Retrieve the attributes for this file, relative to a base directory.
   # Note that Puppet::FileSystem.stat(path) raises Errno::ENOENT
   # if the file is absent and this method does not catch that exception.
-  def collect(source_permissions = nil)
+  def collect
```
Comment (joshcooper) [inline lib/puppet/file_serving/metadata.rb]: same breaking change comment here

File: lib/puppet/file_serving/metadata.rb (hunk 3 above)
Comment (joshcooper) [inline lib/puppet/file_serving/metadata.rb]: I would expect a `puppet agent` run to serialize/deserialize the `source_permissions` option as a String, and then call the setter with a string. It may be that `puppet apply` or running via tests is calling with a symbol? I'm thinking we should be consistent about what we store in the instance variable, so that the comparison below `source_permissions == :ignore` work as expected.

**Comment by whopper** [comment]
As far as I can tell, acceptance looks to be in a good state with this PR. It should now be ready to go pending any further review.

File: acceptance/tests/pluginsync/4420_pluginfacts_should_be_resolvable_on_agent.rb [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    testfile = File.join(tmpdir, 'testfile')
+
+    teardown do
+      on(master, "rm -rf #{codedir}")
+      on(agent, "rm -rf #{factsd}")
+    end
+
+    step "Pluginsync the external fact to the agent and ensure it resolves correctly"
+    on(agent, puppet('agent', "-t --server #{master} --pluginfactdest #{factsd}"), :acceptable_exit_codes => [2])
+      assert_match(/foo is bar/, stdout)
```
Comment (MikaelSmith) [inline acceptance/tests/pluginsync/4420_pluginfacts_should_be_resolvable_on_agent.rb]: Why the extra space here?

Comment (whopper) [inline acceptance/tests/pluginsync/4420_pluginfacts_should_be_resolvable_on_agent.rb]: @MikaelSmith ah, that's just a bit of style I picked up from a few other tests a while back, presumably to make it clear that this assert is connected to the command run above. I can definitely de-indent it.

File: lib/puppet/file_serving/content.rb (hunk 6 above)
Comment (MikaelSmith) [inline lib/puppet/file_serving/content.rb]: Are you proposing to revert the changes to `metadata.rb` and most of the changes to `terminus_helper.rb`? I'm not sure `Puppet::FileServing::Metadata#collect` and `Puppet::FileServing::Content#collect` are a public interface we need to preserve; they're not even part of `Puppet::FileServing::Base`. Putting this in `4.1.0` seems reasonable to me.

File: acceptance/tests/pluginsync/4420_pluginfacts_should_be_resolvable_on_agent.rb (hunk 8 above)
Comment (MikaelSmith) [inline acceptance/tests/pluginsync/4420_pluginfacts_should_be_resolvable_on_agent.rb]: I hadn't seen it before, and in other languages it implies actual control-flow. Just confused me, if we do it in other tests I don't need it changed.

File: lib/puppet/file_serving/content.rb (hunk 6 above)
Comment (whopper) [inline lib/puppet/file_serving/content.rb]: @MikaelSmith: in reference to removing the default argument here, @joshcooper mentioned he was on the fence about doing it. It'd be easy enough to remove it again (might require a spec update). @joshcooper thoughts?

File: acceptance/tests/pluginsync/4420_pluginfacts_should_be_resolvable_on_agent.rb (hunk 8 above)
Comment (whopper) [inline acceptance/tests/pluginsync/4420_pluginfacts_should_be_resolvable_on_agent.rb]: @MikaelSmith a cursory glance around other tests tells me I might just be crazy and may have made that up myself :). I'll unindent them.

File: lib/puppet/file_serving/content.rb (hunk 6 above)
Comment (MikaelSmith) [inline lib/puppet/file_serving/content.rb]: Oh, I missed that you put it back.

Comment (whopper) [inline lib/puppet/file_serving/content.rb]: Aha, I see! I was a bit confused there :)

**Comment by joshcooper** [comment]
@whopper overall looks good.

One thing I noticed is that we are not including the `source_permission` when serializing/deserializing the metadata in https://github.com/puppetlabs/puppet/blob/master/lib/puppet/file_serving/metadata.rb#L107-L136. This doesn't affect the overall functionality, but it could result in inconsistent state. For example, during a `pluginfacts` request, the agent requests `source_permissions=use`, and the server will collect metadata including stat'ing the file owner/group/mode. However, the server will call `Metadata#to_data_hash` to serialize the request (omitting `source_permissions` in the process).

The agent will use `Metadata.from_data_hash` to deserialize the request, but since `source_permissions` aren't specified, the state of the `@source_permissions` field will depend on the default (for that version of puppet agent).

It would probably be better to include `source_permissions` in serialization, so that the state of the Metadata object matches the various inputs that went into their collection.

**Comment by joshcooper** [comment]
When sourcing pluginfacts, `use` and `use_when_creating` return the source owner, group, and mode `0755` (in decimal `493`) And, `ignore` and the default return `Process.euid`, `Process.egid`, `0644` (in decimal `420`). When an invalid permission is specified the server generates an HTTP 400.

```
0 ~/work/puppet (pr/3883)  $ curl -sk -E ./arcturus.p12:password -H "Accept: pson" 'https://arcturus:8140/puppet/v3/file_metadatas/pluginfacts?environment=production&links=follow&recurse=true&ignore=.svn&ignore=CVS&ignore=.git&checksum_type=sha256&source_permissions=use' | jq 'map(.relative_path, .owner, .group, .mode)'
[
  ".",
  0,
  0,
  493,
  "bash.sh",
  0,
  0,
  493,
  "text.txt",
  0,
  0,
  420
]
0 ~/work/puppet (pr/3883)  $ curl -sk -E ./arcturus.p12:password -H "Accept: pson" 'https://arcturus:8140/puppet/v3/file_metadatas/pluginfacts?environment=production&links=follow&recurse=true&ignore=.svn&ignore=CVS&ignore=.git&checksum_type=sha256&source_permissions=use_when_creating' | jq 'map(.relative_path, .owner, .group, .mode)'
[
  ".",
  0,
  0,
  493,
  "bash.sh",
  0,
  0,
  493,
  "text.txt",
  0,
  0,
  420
]
0 ~/work/puppet (pr/3883)  $ curl -sk -E ./arcturus.p12:password -H "Accept: pson" 'https://arcturus:8140/puppet/v3/file_metadatas/pluginfacts?environment=production&links=follow&recurse=true&ignore=.svn&ignore=CVS&ignore=.git&checksum_type=sha256&source_permissions=ignore' | jq 'map(.relative_path, .owner, .group, .mode)'
[
  ".",
  501,
  20,
  420,
  "bash.sh",
  501,
  20,
  420,
  "text.txt",
  501,
  20,
  420
]
0 ~/work/puppet (pr/3883)  $ curl -sk -E ./arcturus.p12:password -H "Accept: pson" 'https://arcturus:8140/puppet/v3/file_metadatas/pluginfacts?environment=production&links=follow&recurse=true&ignore=.svn&ignore=CVS&ignore=.git&checksum_type=sha256' | jq 'map(.relative_path, .owner, .group, .mode)'
[
  ".",
  501,
  20,
  420,
  "bash.sh",
  501,
  20,
  420,
  "text.txt",
  501,
  20,
  420
]
0 ~/work/puppet (pr/3883)  $ curl -sk -E ./arcturus.p12:password -H "Accept: pson" 'https://arcturus:8140/puppet/v3/file_metadatas/pluginfacts?environment=production&links=follow&recurse=true&ignore=.svn&ignore=CVS&ignore=.git&checksum_type=sha256&source_permissions=foobar' -w '\n%{http_code}\n'
Unsupported source_permission foobar
400
```

**Comment by joshcooper** [comment]
When sourcing a file (with a puppet URI), I see similar behavior, with mode 0777 (or 511 in decimal):

```
0 ~/work/puppet (pr/3883)  $ curl -sk -E ./arcturus.p12:password -H "Accept: pson" 'https://arcturus:8140/puppet/v3/file_metadata/modules/mymodule/somecontent.txt?environment=production&links=manage&checksum_type=md5&source_permissions=use' | jq '{path, owner, group, mode}'
{
  "path": "/Users/josh/test/master/source_perms/code/environments/production/modules/mymodule/files/somecontent.txt",
  "owner": 0,
  "group": 0,
  "mode": 511
}
0 ~/work/puppet (pr/3883)  $ curl -sk -E ./arcturus.p12:password -H "Accept: pson" 'https://arcturus:8140/puppet/v3/file_metadata/modules/mymodule/somecontent.txt?environment=production&links=manage&checksum_type=md5&source_permissions=use_when_creating' | jq '{path, owner, group, mode}'
{
  "path": "/Users/josh/test/master/source_perms/code/environments/production/modules/mymodule/files/somecontent.txt",
  "owner": 0,
  "group": 0,
  "mode": 511
}
0 ~/work/puppet (pr/3883)  $ curl -sk -E ./arcturus.p12:password -H "Accept: pson" 'https://arcturus:8140/puppet/v3/file_metadata/modules/mymodule/somecontent.txt?environment=production&links=manage&checksum_type=md5&source_permissions=ignore' | jq '{path, owner, group, mode}'
{
  "path": "/Users/josh/test/master/source_perms/code/environments/production/modules/mymodule/files/somecontent.txt",
  "owner": 501,
  "group": 20,
  "mode": 420
}
0 ~/work/puppet (pr/3883)  $ curl -sk -E ./arcturus.p12:password -H "Accept: pson" 'https://arcturus:8140/puppet/v3/file_metadata/modules/mymodule/somecontent.txt?environment=production&links=manage&checksum_type=md5' | jq '{path, owner, group, mode}'{
  "path": "/Users/josh/test/master/source_perms/code/environments/production/modules/mymodule/files/somecontent.txt",
  "owner": 501,
  "group": 20,
  "mode": 420
}
0 ~/work/puppet (pr/3883)  $ curl -k -E ./arcturus.p12:password -H "Accept: pson" 'https://arcturus:8140/puppet/v3/file_metadata/modules/mymodule/somecontent.txt?environment=production&links=manage&checksum_type=md5&source_permissions=foobar' -w "\n%{http_code}\n"
Unsupported source_permission foobar
400
```

**Comment by whopper** [comment]
@joshcooper makes sense. I'll get this updated as soon as I finish up another (unrelated) PR

**Comment by whopper** [comment]
@joshcooper pushed your suggested change but I won't have a chance to test it / see if there are other needed changes until tomorrow D:

**Comment by whopper** [comment]
@joshcooper update: ran this through acceptance and some manual tests, and everything looks to be in order. Also updated a few specs and the metadata schema.

**Comment by whopper** [comment]
@joshcooper doh, not sure how I missed that :P. Updated

File: lib/puppet/file_serving/metadata.rb [hunk 9]
```diff
@@ -128,9 +135,9 @@ def to_data_hash
           'type'   => checksum_type,
           'value'  => checksum
         },
-        'type'         => ftype,
-        'destination'  => destination,
-
+        'type'               => ftype,
+        'destination'        => destination,
+        'source_permissions' => source_permissions,
```
Comment (joshcooper) [inline lib/puppet/file_serving/metadata.rb]: @nwolfe previously, puppet's ruby implementation of the `file_metadata` endpoint did not include `source_pemissions` when serializing the response. This lead to inconsistent state on the agent if the agent requested `source_permissions` that were different from its default. For example, a 4.0 agent could request `use`, but the deserialized Metadata objects would have an unspecified `source_permissions` value, and the default value in 4 is `ignore`.

We've decided to include `source_permissions` in the response, similar to `checksum_type`. This should not affect compatibility, because the agent doesn't require or use the `source_permissions` value of objects it gets back. But we likely should modify the clojure endpoint to be consistent. Should I file a SERVER ticket?

**Comment by whopper** [comment]
@joshcooper @kylog Added `source_permissions` to the list of required properties, and sadly blew away your comments in the process :(. For posterity:

Josh:
kylog we're modifying the file_metadata rest endpoint to include the source_permissions that were "used" to generate the Metadata object. Do we also need to include source_permissions in the list of requiredProperties? I'm not super familiar with the api schemas.

I also don't want to break compatibility, e.g. a 4.0.0 agent should be able to talk to 4.1.0 server and vice-versa. A 4.1.0 server (including this info) will not break older clients, because old clients don't try to read in the property, and ignore values they don't recognize.

A 4.1.0 client talking to 4.0.0 server will try to read the value, but default to ignore if it is not specified in the serialized data hash.

That said, I'll definitely try that out to make sure.

Kylo:

Ah, yes, good catch. I would add source_permissions to the list of required properties, since (with this PR in place) it would be an error if the generated json did not include source_permissions. Adding it to required ensures that the specs that check against the schema, e.g. https://github.com/puppetlabs/puppet/blob/e3d4a73c4e950452ed532e779e1094aa8ad54068/spec/unit/file_serving/metadata_spec.rb#L135-L137, will enforce that behavior.

Independent of that, I agree on doing some manual verification for both compatibility directions. Strictly speaking, we require updating the server first, but as a matter of practice many people don't, so it'd be nice not to introduce a newer-client-older-server incompatibility if we don't have to.

**Comment by kylog** [comment]
<shakes fist at github comment support>

File: lib/puppet/file_serving/metadata.rb (hunk 9 above)
Comment (nwolfe) [inline lib/puppet/file_serving/metadata.rb]: @joshcooper Thanks for the explanation, that fix sounds kosher.

> But we likely should modify the clojure endpoint to be consistent

We don't yet have a Clojure endpoint for this, we just forward the request into Puppet via JRuby.  I suspect that upgrading our Puppet dependency will be enough for Puppet Server.

> Should I file a SERVER ticket?

Yes, please!  I don't expect much, if any, work to come out of this, but we do have some acceptance test(s) that we can clean up (un-skip) once this fix comes in so it would be good to have a ticket for it.

**Comment by joshcooper** [comment]
Unfortunately this breaks 4.0.0 agents talking to 4.1.0 masters. The agent tries to deserialize the response, and because of the MethodHelper tries to call the `Metadata#source_permissions=` method, which doesn't exist in 4.0.0...

```
[root@hwvft1j1a4531gm environments]# puppet agent -td --server arcturus.delivery.puppetlabs.net --trace
...
Error: /File[/opt/puppetlabs/puppet/cache/lib]: Failed to generate additional resources using 'eval_generate': Could not intern_multiple from text/pson: Invalid parameter source_permissions to object class Puppet::FileServing::Metadata
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/methodhelper.rb:16:in `block in set_options'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/methodhelper.rb:11:in `each'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/methodhelper.rb:11:in `set_options'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/file_serving/base.rb:41:in `initialize'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/file_serving/metadata.rb:118:in `initialize'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/file_serving/metadata.rb:139:in `new'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/file_serving/metadata.rb:139:in `from_data_hash'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/network/formats.rb:110:in `data_to_instance'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/network/formats.rb:94:in `block in intern_multiple'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/network/formats.rb:93:in `collect'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/network/formats.rb:93:in `intern_multiple'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/network/format_support.rb:18:in `convert_from_multiple'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/indirector/rest.rb:237:in `deserialize_search'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/indirector/rest.rb:147:in `search'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/indirector/file_metadata/selector.rb:19:in `search'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/indirector/indirection.rb:260:in `search'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type/file.rb:666:in `perform_recursion'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type/file.rb:635:in `block in recurse_remote'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type/file.rb:634:in `collect'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type/file.rb:634:in `recurse_remote'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type/file.rb:562:in `recurse'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type/file.rb:463:in `eval_generate'
```

**Comment by whopper** [comment]
@joshcooper well shoot.. shall we sync up tomorrow and figure out how to work around this?

**Comment by joshcooper** [comment]
So it looks like the server can't include `source_permissions` without breaking existing clients. Although `Puppet::FileServing::Metadata.from_data_hash` ignores keys it doesn't know about, it calls the super method `Puppet::FileServing::Base.from_data_hash` which relies on `MethodHelper` (and will blindly call setters for each key-value in the hash). And if the class doesn't respond to a particular setter, it raises an exception: `raise ArgumentError, "Invalid parameter #{param} to object class #{self.class}"`

So I'm thinking we should:

1. Drop commit 6a5e81d74 and merge this PR.
2. File a new ticket for 4.1.0 to future-proof `Base.from_data_hash` so it will ignore parameters it doesn't know about. 
3. File a new ticket for 5.0 to essentially add commit 6a5e81d74 back in.

We could also consider modifying the two places where we call `Metadata.indirection.{find,search}` and explicitly set `source_permissions=` on the returned objects (similar to how we call `Metadata#source=`)? On one hand it would ensure the returned Metadata objects are in a consistent state. But on the other hand, it feels like a bandaid that would only confuse things (some data is specified on the server, other data is specified on the client). And if intend to support 4.x agents talking to 4.x and 5.x masters, things get really confusing.

**Comment by whopper** [comment]
@joshcooper I went ahead and dropped 6a5e81d74 for the time being.

**Comment by MikaelSmith** [comment]
Merged #3933, can you rebase and verify the acceptance test works?

**Comment by whopper** [comment]
@MikaelSmith rebased, and the test indeed passes (also verified that it failed *before* rebasing)
